### ISSUE TITLE
Move Trainer's tokens per second into the inner training loop

### ISSUE THREAD

**Issue by crclark** [huggingface/transformers#34471]
### Feature request

Right now `include_tokens_per_second=True` in `Trainer` only reports the tokens per second metric at [the end of training](https://github.com/huggingface/transformers/blob/c1753436dbb8bcbcee183cdd6eba9f08a90d602a/src/transformers/trainer.py#L2610). It would be very useful to have this metric reported continuously inside the training loop so we can monitor it during training. 

### Motivation

The current behavior is counter-intuitive, doesn't align with other convenient trainers (like torchtune), and it's undocumented, so I had to RTFC to figure out why the metric wasn't showing up.

### Your contribution

I probably don't have time to contribute it myself.

**Comment by Rocketknight1** [comment]
cc @muellerzr @SunMarc
